### PR TITLE
added ado alerts doc

### PR DESCRIPTION
**Pull request by vinaybommana**
this is regarding https://issues.redhat.com/browse/MTSRE-1053 to document the steps and the process for defining an alert for ADO metrics in RHOBS osd tenant

cc @pepedocs

### CODE REVIEW

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+---
+title: Adding ADO alerts
+linkTitle: Adding ADO alerts
+---
+
+This guide describes how to add alerts for addon-operator in OSD RHOBS tenant
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
This guide describes how to add alerts for addon-operator in the OSD RHOBS tenant.
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+## Background
+
+Since the addon-operator doesn’t have a specific tenant in RHOBS with a service account,
+the addon-operator metrics are scraped to OSD tenant with its own service account,
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
the addon-operator metrics are scraped to the OSD tenant with its own service account,
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+so instead of adding rules using [obsctl cli](https://github.com/observatorium/obsctl) and
+syncing them using [obsctl-reloader](https://github.com/rhobs/obsctl-reloader)
+SRE-P has a repo called [rhobs-rules-and-dashboards](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards)
+which based on the tenant will automatically sync rules defined in the repo to app-interface.
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
which, based on the tenant, will automatically sync rules defined in the repo to app-interface.
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+## Overview
+
+we need to have the following prerequisites:
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
We need to have the following prerequisites:
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+If the tenant is not defined in the repo, we need to follow the process defined
+[here](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/rules#registering-a-new-tenant)
+which explains how to register the tenant in the app-interface and how to define obsctl-reloader
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
which explains how to register the tenant in app-interface and how to configure obsctl-reloader
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+to sync rules for the particular tenant.
+
+Since osd tenant is already registered and the obsctl-reloader configuration is already
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
Since the OSD tenant is already registered and the obsctl-reloader configuration is already
```

Comment (garrettthomaskth) [reply content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]: Should "Since" be in this sentence?

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+defined [here](https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/data/services/rhobs/observatorium-mst/cicd/saas.yaml#L127)
+(we should see osd in **MANAGED_TENANTS** parameter in the **observatorium-mst-common** named
+item and secrets are defined [here](https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/data/services/rhobs/observatorium-mst/namespaces/app-sre-stage-01/observatorium-mst-stage.yml#L79))
+
+## Steps
+
+we can define rules for the addon-operator metrics in the [rhobs-rules-and-dashboards](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards)
+repo in the [rules/osd](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/rules/osd)
+folder (the tenants are added as specific folders) and tests for the prometheus rules are
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
folder (the tenants are added as individual folders) and tests for the prometheus rules are
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+defined in the [tests/rules/osd](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/test/rules/osd)
+folder.
+The suggested way to add folder names and rules filenames for an alert is explained [here](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/rules#one-folder-per-targeted-rhobs-tenant).
+
+tldr: create file name with `.prometheusrule.yaml` as suffix at the end,
+add tenant label for the `PrometheusRule` object.
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
add tenant label to the `PrometheusRule` object.
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+item and secrets are defined [here](https://gitlab.cee.redhat.com/service/app-interface/-/blob/master/data/services/rhobs/observatorium-mst/namespaces/app-sre-stage-01/observatorium-mst-stage.yml#L79))
+
+## Steps
+
+we can define rules for the addon-operator metrics in the [rhobs-rules-and-dashboards](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards)
+repo in the [rules/osd](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/rules/osd)
+folder (the tenants are added as specific folders) and tests for the prometheus rules are
+defined in the [tests/rules/osd](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/test/rules/osd)
+folder.
+The suggested way to add folder names and rules filenames for an alert is explained [here](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/rules#one-folder-per-targeted-rhobs-tenant).
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
The suggested file naming conventions are described [here](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/rules#one-folder-per-targeted-rhobs-tenant).
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+tldr: create file name with `.prometheusrule.yaml` as suffix at the end,
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
tldr: create a file name with `.prometheusrule.yaml` as a suffix,
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+add tenant label for the `PrometheusRule` object.
+
+### Adding promrules
+
+For creating a PoC alert we have selected `addon_operator_addon_health_info` metric
+since it basically explains the addon health for particular `version` and `cluster_id`.
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
since it basically explains the addon health for a particular `version` and `cluster_id`.
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+defined in the [tests/rules/osd](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/test/rules/osd)
+folder.
+The suggested way to add folder names and rules filenames for an alert is explained [here](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/tree/main/rules#one-folder-per-targeted-rhobs-tenant).
+
+tldr: create file name with `.prometheusrule.yaml` as suffix at the end,
+add tenant label for the `PrometheusRule` object.
+
+### Adding promrules
+
+For creating a PoC alert we have selected `addon_operator_addon_health_info` metric
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
For creating a PoC alert we have selected the `addon_operator_addon_health_info` metric
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+since it basically explains the addon health for particular `version` and `cluster_id`.
+The metrics for creating alerts can be decided and the promql queries
+can be tested out from [promlens stage](https://promlens.stage.devshift.net)
+or [promlens prod](https://promlens.devshift.net).
+The metrics are scraped from addon-operator to
+observatorium-mst-stage (stage) / observatorium-mst-prod (prod) datasource.
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
the observatorium-mst-stage (stage) / observatorium-mst-prod (prod) datasources.
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+Sample `addon_operator_addon_health_info` metric data
+
+```yaml
+addon_operator_addon_health_info{_id="08d94ae0-a943-47ea-ac29-6cf65284aeba", container="metrics-relay-server", endpoint="https", instance="10.129.2.11:8443", job="addon-operator-metrics", name="managed-odh", namespace="openshift-addon-operator", pod="addon-operator-manager-7c9df45684-86mh4", prometheus="openshift-monitoring/k8s", receive="true", service="addon-operator-metrics", tenant_id="770c1124-6ae8-4324-a9d4-9ce08590094b", version="0.0.0"}
+```
+
+The particular metric gives information about
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
This particular metric gives information about
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+For creating a PoC alert we have selected `addon_operator_addon_health_info` metric
+since it basically explains the addon health for particular `version` and `cluster_id`.
+The metrics for creating alerts can be decided and the promql queries
+can be tested out from [promlens stage](https://promlens.stage.devshift.net)
+or [promlens prod](https://promlens.devshift.net).
+The metrics are scraped from addon-operator to
+observatorium-mst-stage (stage) / observatorium-mst-prod (prod) datasource.
+
+Sample `addon_operator_addon_health_info` metric data
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
Sample `addon_operator_addon_health_info` metric data:
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+```
+
+The particular metric gives information about
+`version`, `cluster_id (_id)` and `addon_name (name)` which can be used to create the alert.
+
+- We should ignore the metrics with version `"0.0.0"`
+- The addon health is `"Unhealthy"` if no version exists for the particular
+addon with [value 1](https://github.com/openshift/addon-operator#monitoring-and-metrics)
+- In theory, the latest version for the particular addon
+should have value of 1 if not then the addon is `"Unhealthy"`
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
should have the value of 1. If not, then the addon is `"Unhealthy"`
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+
+The prometheus rule for the addon_operator_addon_health_info metric is defined [here](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/blob/main/rules/osd/addon-operator-addon-health-info-rules.prometheusrule.yaml)
+
+```yaml
+expr: (count by (name, _id) (addon_operator_addon_health_info{version!="0.0.0"})) - (count by (name,_id) (addon_operator_addon_health_info{version!="0.0.0"} == 0)) == 0
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md:71]: Is this expression the same as
```
(count by (name,_id) (addon_operator_addon_health_info{version!="0.0.0"} != 0)) == 0
```
?

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+aggregating the metrics using `name, _id`
+we need to raise an alert for the metric with an `id, name` which does not have at least
+one value of 1 in the aggregated metrics for the particular name, _id
+
+from the entire addon_operator_addon_health_info metrics aggregated by count
+`[(count by (name, _id) (addon_operator_addon_health_info{version!="0.0.0"}))]`
+subtract the metrics having 0 values
+`[(count by (name, _id) (addon_operator_addon_health_info{version!="0.0.0"} == 0))]`
+if the result is `0` value that means the metric does not have any other values except `0`
+raise the alert
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
We aggregate all metrics on `name` and `_id` and count the non-0 value metrics. If the count is 0 (implying there has been no non-0 value metrics), raise the alert.
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+### Writing tests for the promrules
+
+Create a `<alertname>.promrulestests.yaml` in the test/rules/<tenant>/ folder,
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
First create a `<alertname>.promrulestests.yaml` file in the `test/rules/<tenant>/` folder.
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+it is advised to test out different scenarios for the edge cases so
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
It is advised to test out different scenarios for the edge cases so
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+that the expr defined in the rules/<tenant> folder will raise alert if say the addon is `"Unhealthy"`.
+Try out different scenarios by defining different series as
+input to the test rules such as [here](https://gitlab.cee.redhat.com/service/rhobs-rules-and-dashboards/-/blob/main/test/rules/osd/addon-operator-addon-health-info-rules.prometheusrulestests.yaml#L8).
+
+The  tests can be validated by running: `make test-rules`
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
The tests can be validated by running: `make test-rules`
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+in the rhobs-rules-and-dashboards directory.
+NOTE: Make sure that the tests are defined in the tests/rules/<tenant> folder
+
+Since the pagerduty config is defined per tenant basis which is `osd tenant` in this case,
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]:
```suggestion
Since the pagerduty config is defined on a per tenant basis which is `osd tenant` in this case,
```

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+the alert will be triggered and paged to SRE-P,
+so a proper runbook should be added redirecting the alert to lp-sre folks.
```
Comment (garrettthomaskth) [inline content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md:93]: Is this a TODO?

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md (hunk 17 above)
Comment (vinaybommana) [reply content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md:71]: so the promql won't return anything if `addon_operator_addon_health_info{version!="0.0.0"} != 0` doesn't return any hits. so subtracting 0 value metrics from everything is the way

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md (hunk 23 above)
Comment (vinaybommana) [reply content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md:93]: no we have already added the runbook for the alert

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md (hunk 6 above)
Comment (vinaybommana) [reply content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md]: I don't think we should have Since there, removing it now

File: content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md (hunk 17 above)
Comment (vinaybommana) [reply content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md:71]: I tested this out, in the prom tests

Comment (garrettthomaskth) [reply content/en/docs/internal-documentation/tutorials/adding_ado_alerts.md:71]: Ohhh yeah I understand :) Okay cool!
